### ISSUE TITLE
Flag issues: doesn't work in PMs, needs two flags before they can be cleared, etc

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#4648]
Some issues with the feature for reporting violations of the community guidelines:
- The flag icon does not work in personal messages. Clicking on it produces no result (no popup, no error, nothing else). There are no relevant messages in Chrome's JavaScript Console either when loading the inbox page or when clicking the icon.
- EDIT: I've fixed this: Inappropriate flags cannot be cleared by a moderator until there are two flags. (Workaround is for the moderator to flag the message themselves then reload and clear the flags.)

And for discussion:

This was posted to a guild a few days ago (sorry, I should have passed it on sooner):
_"I'm finding the report a post message intimidating. It's literally scaring me off from reporting posts, since who knows if I'm reporting a post that is actually fine? Or it's borderline? Or something? It makes me scared to be wrong, and in this case, I think it's better to be a little wrong than never reported at all."_
Perhaps we could include a short list of common things that are appropriate for reporting? Something like, "... including but not limited to: swearing, posting multiple nonsensical messages, bigotry, ..." These things are all in the Guidelines of course, but you know what people are like about reading them. :)

Also for discussion: 

I would REALLY like to be able to flag my own messages. Pretty please! :) It would make testing the flag system easier - no more messing about with two accounts. There's been three or four occasions now when that would have been useful to me, and I've seen one other dev make a comment about how he wasn't able to test something because of the limitation. I can even see a reason for it to be a useful feature: imagine your young sibling gets onto your computer and posts something inappropriate using your account. You see the post but not until someone else has already flagged it once. You are appalled that your account has been used like this and worried about getting in trouble, and so you flag your "own" message yourself to show contrition. Okay, it's a bit of a stretch, but I can imagine it happening. :) I also can't see any reason to not allow people to flag their own messages; they aren't likely to do it accidentally.

And just a note: @crookedneighbor I'm LOVING the flag system. It makes moderating so much easier! Not just because non-moderators can help more now and because we are rapidly notified of bad posts, but also because there's no longer any need to make screenshots of posts to show Leslie before we delete them. It's such a nice convenience! :)
##

**Comment by lemoness** [comment]
Agreed, the flag system is such a relief. It's made me so much calmer about
leaving the Tavern unmonitored for a stretch of time, because inappropriate spots get flagged very quickly.

I'd be happy to include that list of examples, and I agree that the ability
to flag your own message is a good idea.

**Comment by crookedneighbor** [comment]
Figured out why the flagging modal doesn't open up, it's because function for opening it was inside the chat controller instead of the general purpose group controller. 

While the modal now opens up in my branch to fix this, the report abuse button displays an error when clicked. The problem is that the flagging api is set up for groups, while the private messaging is through members, so it can't look up the chat message since it's not located in the groups collection.

**Comment by Alys** [comment]
Summary of actions required (they don't all have to be done at once - e.g., it's okay to do just one and leave the others for someone else to do):
1. allow flagging of private messages
2. allow a user to flag their own messages
3. add additional text to the flag confirmation screen to give a better idea of what should be flagged. Probably one extra sentence and a few bullet points of examples. If you're interested in doing this, comment here and we'll prepare some text for it.
4. put a message at the top of the inbox to say that if you've received an inappropriate PM, you should email a screenshot of it to Lemoness (leslie@habitica.com)  [_EDIT: I added this after Lemoness gave her approval for it below_]

**IMPORTANT** If point 1 requires changes to the API it should not be done now because API v3 is being prepared, and so v2 shouldn't be modified unless essential. Also, if it requires significant changes to other sections of the code, it migh alsot be best to not do it now because our senior devs are busy with the new API and might not be available to do a code review.

As a temporary workaround to point 1, we could put a message at the top of the inbox to say that if you've received an inappropriate PM, you should email a screenshot of it to Lemoness (leslie@habitica.com) -- @lemoness, what do you think of this idea?

**Comment by lemoness** [comment]
I think that is a good workaround until we have flagging.

**Comment by pnrich** [comment]
I've looked into this. Assuming I have the right idea of what constitutes "Changing the API", the following is true. I assume any files under directories such as /api-v2 are part of the API and shouldn't be changed. Correct me here if I'm wrong.

Point 1 above is impossible without changing the API. The flagging of private messages would require the ability to set a flag without specifying a groupId. Currently, the API sets flags via a POST to /api/v2/groups/{groupId}/chat/{messageId}/flag. See public/js/services/chatServices.js, line 12. That URL obviously requires a groupId.

Point 2 would also require a change to the API. Currently, if the reporter user id is the same as the message author user id, then the API sends a 401 response and an error message. See src/controllers/api-v2/groups.js, line 345.

Point 3 is simple. It would just require modifying the text for the abuseFlagModalBody tag. See locales/en/groups.json, line 104. Once that's done, it's just a matter of translating it.

Point 4 is similarly easy. Just modify the text for the blurbInbox tag. See locales/en/contrib.json, line 59.

TL;DR Points 1 and 2 require changing the API. I could easily make the changes for points 3 and 4.

**Comment by Alys** [comment]
@pnrich I agree with your assessment. If you'd like to work on points 3 and 4, that would be great. 

For point 3, I'll prepare an initial body of text and ask Lemoness to make changes, and then you could implement it - that would save you time over adding text now and then having to make changes. :) I'll get that done early this weekend if I can!

**Comment by pnrich** [comment]
Sounds great!

**Comment by Alys** [comment]
@lemoness, what do you think of this for the flag confirmation dialog:

Current text:

_Are you sure you want to report this post? You should ONLY report a post that violates the [Community Guidelines](https://habitica.com/static/community-guidelines) and/or [Terms of Service](https://habitica.com/static/terms). Inappropriately reporting a post is a violation of the Community Guidelines and may give you an infraction._

Proposed text (the first three sentences are unchanged):

---

**Are you sure you want to report this post? You should ONLY report a post that violates the [Community Guidelines](https://habitica.com/static/community-guidelines) and/or [Terms of Service](https://habitica.com/static/terms). Inappropriately reporting a post is a violation of the Community Guidelines and may give you an infraction. Appropriate reasons to flag a post include but are not limited to:**
- **swearing, religous oaths**
- **bigotry, slurs**
- **adult topics**
- **violence, including as a joke**
- **spam, nonsensical messages**

---

Current screen:

![screen shot 2015-12-20 at 4 14 38 pm](https://cloud.githubusercontent.com/assets/1495809/11916821/e6e4bb1a-a734-11e5-8ea2-6f439d5dcc86.png)

**Comment by lemoness** [comment]
Looks great!

**Comment by pnrich** [comment]
I've almost finished this, but one part has really got me stuck. I'm trying to have the text from part 4 (modifying blurbInbox) include a mailto link to leslie@habitica.com. However, it's not accepting it as a link, and in my tests, the whole <a> tag shows up instead of a link. This is weird, because links appear elsewhere in translated strings, and function properly, for example in groups.json:

`"inviteAlertInfo": "If you have friends already using Habitica, invite them by <a href='http://habitica.wikia.com/wiki/API_Options' target='_blank'>User ID</a> here."`

There are even mailto links elsewhere in the translations, in communityGuidelines.json

`please email Lemoness (<a href=\"mailto:leslie@habitica.com\">leslie@habitica.com</a>)`

I tried both of these formats, and neither worked for blurbInbox. Is seems like that string is handled differently, such that it won't correctly format a link. Please advise.

**Comment by Alys** [comment]
@pnrich I think using `!=env.t('....')` rather than `=env.t('....')` (note the ! mark) will help with that.

**Comment by pnrich** [comment]
That did the trick. Creating a pull request soon.

**Comment by librarianmage** [comment]
What is left of this Issue is

1. allow flagging of private messages
2. allow a user to flag their own messages

right?

**Comment by SabreCat** [comment]
Client UI issues. Pending for after frontend redesign.

**Comment by lemoness** [comment]
So, users should already be able to flag their own posts. Should we instead just close this and make a separate ticket for allowing people to flag private messages?

**Comment by Alys** [comment]
Yes, good idea, this issue is too full of irrelevancies now. I'm creating a new issue. https://github.com/HabitRPG/habitica/issues/9204
